### DISCUSSION TITLE
Model partially updated after one single transaction seemingly went fine.

### DISCUSSION THREAD

**Discussion by Joacchim** [sqlalchemy/sqlalchemy#10503] [Usage Questions] [answered]
Hello !

First of all, thanks for the wonderful work on sqlalchemy ! :)

For a bit of context first:
I've started using sqlalchemy a few months back for a work project, essentially a scheduler or sorts, with multiple services interacting, to notify each other of available work, or completed work (that may have freed up resources). The core service of the said scheduler is the only one with the logic to update one specific model, representing a portion of a task (namely a `Job`).

Now, I've been encountering a bizarre situation:
When this core service is required to handle completion events for the ongoing `Jobs`, and update the model objects according to the result (success, failure, etc); the logs indicate a complete (and seemingly correct) handling of the model, and its relationships.
Then, at the following round of checks, one of these very `Job` model objects happens to appear partially updated, breaking the assumptions that I made about its data/state upon acknowledging a work completion.

I essentially trusted the transaction logic to guarantee an all-or-nothing update, but somehow, I get a partial object, giving me what I'd consider to be a "corrupt state" of my model object. Given my current code, I'm hardly considering a mis-use of the transactional capabilities (as all relevant fields of the table/model are updated at once), and can only consider it to be a Bug. That being said, as I'm unable to reproduce it reliably, the issue template seemed to indicate it'd be better to start this off with a discussion, so here I am :)

As this problem does not occur consistently, but "once every now and then", I cannot reproduction case script. I can, at best, try to transcribe a simplified version of my logic, in hopes that it may give rise to insights from the community.

Here are the versions of the relevant dependencies that I can think of
 - Environment/OS: Debian/bullseye 
 - Python: 3.9.2
 - SQLAlchemy: 2.0.19
 - DB Engine: mysql + asyncmy (0.2.8)
 
As for a rough pseudo code of the transactional logic:
```python3
import sqlalchemy as sa
from sqlalchemy import orm 

class Task(BaseModel):
    __tablename__ = "task"
    id = orm.mapped_column(sa.Integer, primary_key=True)
    result = orm.mapped_column(sa.Integer)
    jobs = orm.relationship("Job", back_populates="parent")
    
class Job(BaseModel):
    __tablename__ = "job"
    id = orm.mapped_column(sa.Integer, primary_key=True)
    state = orm.mapped_column(sa.Integer)
    result = orm.mapped_column(sa.Integer)
    attempt = orm.mapped_column(sa.Integer,)
    budget = orm.mapped_column(sa.Integer)
    task_id = sa.Column(sq.Integer, sa.ForeignKey("task.id"))
    parent = orm.relationship("Task", back_populates="jobs")
    

class Work:
    result: int

async def get_completed_work_status(job_id: int) -> Optional[Work]:
    """ Assumed to return an actual work status, skipping actual logic as it is not relevant """

# Completion handling routine
async def handle_completion(transaction, job):
     if (work := await get_completed_work_status(job.id)):
          # About the BUG/Behavior:
          # One or Two of the three columns updated here are not persistently updated?
          job.state = 1 #  equivalent of "done/completed"
          job.result = work.result
          job.attempts += 1
          transaction.add(job)
          # Update parent's result when all jobs are complete
          if all([j.state == 1 for j in job.parent.jobs]):
              job.parent.result = int(all([j.result ==1 for j in job.parent.jobs]))
              transaction.add(job.parent)

# Event callback, called upon work completion
async def completion_cb(job_id: int):
    session = AsyncSession(db_engine, expire_on_commit=False)
    async with session as transaction:
        job = await transaction.scalar(
            sa.select(Job).options(joinedload(Job.parent).joinedload(Task.jobs).where(Job.id == job_id)
        )
        await handle_completion(transaction, job)
        await transaction.commit()
```

**Comment by zzzeek** [comment]
OK, this is the important part:

> I essentially trusted the transaction logic to guarantee an all-or-nothing update, but somehow, I get a partial object, giving me what I'd consider to be a "corrupt state" of my model object. 

That indicates the state on your Python object is wrong - however, it is critical to then stop exactly at that point (using [pdb](https://docs.python.org/3/library/pdb.html) or a very long sleep), assuming you expect all data to be committed, open up a terminal to your actual database and see what the actual state is within the DB.   It's two totally different problems where one is the data is not actually being committed, vs. the object is showing stale database state.

to make sure your object shows everything fresh from the DB connection you can `session.refresh(obj)` it which will refresh attributes; or load it using [populate_existing](https://docs.sqlalchemy.org/en/20/orm/queryguide/api.html#orm-queryguide-populate-existing).

**Reply by Joacchim** [reply]
Sorry, I kinda got lost reading your answer, so I'll try to rephrase it to confirm my understanding:
There are actually two problems occurring here (in all probability):
 1. The commit is actually not happening (I wonder why I don't get any error then)
 2. The "corrupt" state I observe "could" be a stale database state
 
 Where I get lost in this is the following happening:
  - No error from transaction
  - What I consider to be a "corrupt state", is one where part only of the model was updated (in my case, the equivalent of the `attempt` field here). Where does this state come from if not from the DB ?  Because on subsequent "round" of check, I do retrieve the object again from the DB. Is it somehow cached somewhere ?
  - Even despite a full service restart, I still see the "corrupt state" of the object, hinting that it's actually what got written to the DB ?

**Reply by Joacchim** [reply]
Also, as it's a scheduler that's supposed to be running in the background, it's kinda hard to be there to investigate.
  
Now, when should I attempt to start a pdb/whatever ? Because most operations actually go fine, and I got no idea how to determine which is failing.

Except maybe attempting to retrieve the object after the commit, and double-checking data ?

**Reply by zzzeek** [reply]
OK if it's like a background thing, then yes add these assertions to your code.   you could use a regular engine (even in async, it's fine) and just to "with engine.connect() as conn" then run queries on that connection to assert what you expect to see.

additionally you can turn on [sql logging](https://docs.sqlalchemy.org/en/20/core/engines.html#configuring-logging) to see if the expected statements take place.

as a general technique, if there was some version of this code that didn't have a problem that it now has, comparing diffs to see what's changed is helpful too.

**Comment by Joacchim** [comment, marked as answer]
Hello,

Throughout my stabilization phase of the aforementioned service, it seems indeed that I've been overlooking the potential of concurrent operations, which were re-applying the logic on top of already-updated entries.
This was most likely the cause of the unexpected behavior I was observing, though activating SQL logs did not help me identify this; as it's been quite a while since I've last seen the issue.

Thus, I apologize for the inconvenience.
I sincerely thank your for the time you spent answering me, and the explanations you gave about the inner mechanisms !
